single_kernel_mongo/managers/mongo.py

```diff
@@ -0,0 +1,415 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""The Mongo manager.
+
+In this class, we manage the mongo database internals.
+
+This class is in charge of creating users, databases, initialising replicat sets, etc.
+"""
+
+from __future__ import annotations
+
+import json
+import logging
+from typing import TYPE_CHECKING
+
+from dacite import from_dict
+from ops import Object
+from ops.model import Relation
+from pymongo.errors import PyMongoError
+
+from single_kernel_mongo.config.literals import Substrates
+from single_kernel_mongo.core.structured_config import MongoDBRoles
+from single_kernel_mongo.exceptions import DatabaseRequestedHasNotRunYetError, SetPasswordError
+from single_kernel_mongo.lib.charms.data_platform_libs.v0.data_interfaces import (
+    DatabaseProviderData,
+)
+from single_kernel_mongo.managers.k8s import K8sManager
+from single_kernel_mongo.state.charm_state import CharmState
+from single_kernel_mongo.utils.mongo_config import (
+    EMPTY_CONFIGURATION,
+    MongoConfiguration,
+)
+from single_kernel_mongo.utils.mongo_connection import MongoConnection, NotReadyError
+from single_kernel_mongo.utils.mongodb_users import (
+    OPERATOR_ROLE,
+    BackupUser,
+    MongoDBUser,
+    MonitorUser,
+    OperatorUser,
+)
+from single_kernel_mongo.workload.mongodb_workload import MongoDBWorkload
+
+if TYPE_CHECKING:
+    from single_kernel_mongo.core.operator import OperatorProtocol
+
+logger = logging.getLogger(__name__)
+
+
+class MongoManager(Object):
+    """Manager for Mongo related operations."""
+
+    def __init__(
+        self,
+        dependent: OperatorProtocol,
+        workload: MongoDBWorkload,
+        state: CharmState,
+        substrate: Substrates,
+    ) -> None:
+        super().__init__(parent=dependent, key="managers")
+        self.charm = dependent.charm
+        self.workload = workload
+        self.state = state
+        self.substrate = substrate
+        pod_name = self.model.unit.name.replace("/", "-")
+        self.k8s = K8sManager(pod_name, self.model.name)
+
+    def mongod_ready(self, uri: str | None = None) -> bool:
+        """Is MongoDB ready and running?"""
+        config = EMPTY_CONFIGURATION
```

Comment:
should this be a variable

Reply:
I don't know, because I don't see places where we use more than an unauthenticated uri.
From time to time we use a uri which is built on the fly because it does not belong to any user.
What can be done however is to drop the `config =` and use directly the empty configuration in the Mongo Connection.

In the former codebase, we would pass `None` to the `MongoConnection` object but I dropped it considering that we should always provide a configuration even when it's empty so that it makes typing easier and more straightforward.

Reply:
I'm suggesting that on line 73 we change it to 
```
        with MongoConnection(EMPTY_CONFIGURATION, actual_uri, direct=True) as direct_mongo:
```

since `config` isn't actually modified it seems. Sorry I should have made that into a GH suggestion